### ISSUE TITLE
Fairness and freedom for artists: Towards a Robot Economy for the Music (Streaming) Industry

### ISSUE THREAD

**Issue by Tim-W** [Tribler/trustchain-superapp#45]
(Placeholder issue for msc thesis)

**Problem statement**
<IMG src="https://user-images.githubusercontent.com/4957928/74647933-e64fb780-517c-11ea-94d8-c4c6cdb501e0.png" height="422">
(Musicians/artists/creators) receive low compensation for publishing their content. 
* [Typical cut of revenue:](https://www.theguardian.com/technology/2015/apr/03/how-much-musicians-make-spotify-itunes-youtube) (distributor: Spotify/iTunes/Google Play/...)
Signed records: 25% distributor 55% label 20% artist
Unsigned records: 40% distributor 60% artist
* Smaller artists get low revenue, due to the "pro rata" model applied by e.g. Spotify. A [user-centric](https://www.rollingstone.com/music/music-features/should-spotify-change-the-way-it-pays-artists-763986/) model should be more fair (economics/business, less CS)
* [The distributor could censor tracks](https://www.springfieldnewssun.com/entertainment/spotify-removes-kelly-music-from-playlists-under-new-hateful-content-policy/9NIgUD8qJBd2tAYmVhn5hK/)
* [Platform lock-in](https://www.google.com/search?q=Google+Rules%3A+The+History+and+Future+of+Copyright+Under+the+Influence+of+Google)

**Existing and upcoming solutions**
Blockchain-based solutions exist for music streaming services. At least two start-ups attempted.
* [OPUS Audio](https://opus.audio/) "Helps artist generate the revenue they deserve". 
It uses [IPFS](https://ipfs.io/) for distributed file storage, but actually uses a back-up off-chain database for performance purposes. Strategy for IPFS nodes: OPUS deploys **its own** nodes around the globe for low latency. 
OPUS uses Ethereum smart contracts. Using smart contracts, "[consumers] can rest assured almost 100% of the payments are delivered to the artists transparently, immutably and without intermediaries taking large chunks of revenue”.
* [Musicoin project](https://musicoin.org/) 
encourages **independent artists** to register and publish their work on its own blockchain-based platform. Very interesting: consumers have a free service without ads, while artists get paid (in $MUSIC)
Uses earn-per-play smart contracts based on preset fees each time a song gets played (for >s seconds). Consumers can also tip artists (micropayments). 
Uses mining in the form of bitcoin.
* [Mycelia project](http://myceliaformusic.org/) is not a streaming service on its own but does provide "creative passport" technology: a digital container with authorization of work, acknowledgements, payment mechanisms and metadata.

**Shortcomings of these solutions**
* Musicoin pay-per-play system depends entirely on investors and the price of its non-stable currency. If its currency becomes [worthless](https://coinmarketcap.com/currencies/musicoin/) then the artists also receive low revenue.
* OPUS deploys nodes itself and consumer pay a (small) fee for this infrastructure (different from Tribler/torrents)

**The dream: a good decentralized streaming system**
* A (torrent) streaming technology as fast as spotify, based on a p2p file sharing system (tribler? IPFS?). Challenge: seeders for content with small popularity. To think about: are bandwidth tokens enough incentive?
* Proof of authorship on the blockchain Tribler/tribler#4144
* (Near) 100% of fees from consumers go directly to the artists they listen to.
* Payouts with stablecoin (USDT/EURS)
* Tipping system/micropayments
* Effective sharing and recommendation systems

**Completed reading list**
* [How the blockchain is disrupting the art economy as we know it (2017)](https://www.forbes.com/sites/rogeraitken/2017/08/17/how-the-blockchain-is-disrupting-the-art-economy-as-we-know-it/#53c12c9f74fe)
* de Vos, M., & Pouwelse, J. (2018). A Blockchain-based Micro-Economy of Bandwidth Tokens. CompSys 2018.
* [Blockchain is changing how media and entertainment companies compete](https://sloanreview.mit.edu/article/blockchain-is-changing-how-media-and-entertainment-companies-compete/)
* Zhang, B. (2018). Credit Mining: An Incentive and Boosting System in a Peer-to-Peer File-sharing Network.

**Comment by Tim-W** [comment]
Draft version 3 of problem description: https://github.com/Tim-W/msc-thesis/blob/master/report.pdf

Removed some sections about privacy/GDPR/user data and improved focus on economic and power aspects of Spotify.
Based the red line of the problem around 4 pillars: gatekeeping, intermediaries taking large cut, monopsomy, recommendations and curatorial power. Found some articles about Spotify-curated playlists, which was an eye-opener for me. 99 out of top 100 popular playlists on Spotify are Spotify-curated and have large influence. Draws attention away from labels and towards Spotify-signed records in the future (just like Netflix is doing). Also part about human influence on top of 'black box recommendation machine learning' is interesting for my prototype: No artist knows why it is visible or invisible on some playlist, no one knows how to get promoted on playlists, no one knows how recommendations work.

**Comment by devos50** [comment]
Just did a quick read-through of the problem description. A few comments:
- Your research question ("How can we design and implement a music streaming service that distributes the power from one authority to its users?") can be more precise. From what I understand, you are not building a platform around the redistribution power process, but a platform where this empowerment is already built-in. Your research question should reflect this. Furthermore, I think your topic resolve around the concept of *disintermediation*, the process of taking out the middlemen or reducing the extent to which it can exercise its power on the platform (judging from the thesis title). I also suggest evaluating your prototype for more credibility. Proposed research question: "How can we design, implement and evaluate a disintermediated music streaming service?"
- For a better storyline flow, I suggest moving your research question to the end of the problem description, after you introduced the four pillars. These four problems then come together in your research question. I like how these pillars are divided.
- Make sure to give proper citations for the claims you make in the introduction sentences in Section 2.
- It seems you are using subsubsections instead of subsections.

**Comment by Tim-W** [comment]
Thanks for the feedback @devos50 !
> Make sure to give proper citations for the claims you make in the introduction sentences in Section 2.
This is an introduction that uses statements, which I afterwards backup using the sources in the paragraphs below it. Does that make sense? 
Good ideas for the research question, which indeed needs some thought and revision. 

I created various UI and UX improvements for the superapp MusicDAO. Such as: a button in action bar that shows up/down and other connectivity stats, showing loading info, improvements to layout. More detail can be seen in description of this PR: https://github.com/Tribler/trustchain-superapp/pull/46

**Comment by Tim-W** [comment]
@synctext Some questions about the next steps of this thesis: 
I was working on cryptocurrency integration for payment to artists. BTC wallet is already integrated in the superapp, as a light wallet. However I'm thinking about integrating something with a DAG (like NANO) for scalability and fast tx confirmation time for a demo. Does that make sense?

The issue I see with integrating this with our MusicDAO, is that these cryptos are still dependent on external servers (namely high throughput nodes), which makes it not truly autonomous, "phone-to-phone serverless". I don't think there exists a solution for this currently, so it should just be mentioned in future work?

How should the musicDAO decide which crypto **node** to use? I could run my own node and connect the superapp to it, but this moves towards a more centralized architecture. On the other hand I could connect the superapp to an external node, but then we rely on node uptime from an external party. 

So TLDR: 1. is it a good idea to work on crypto payments now? If so, how do we preserve the decentralized, autonomous nature of the MusicDAO?

**Comment by synctext** [comment]
please focus on Bitcoin, Trustchain and music (industry) DAO.

Bitcoin has existing fast confirm ability using server optionally. Would advise to first perfect the music (search)  part of your app before switching to replacing operational Bitcoin libs. Other DAGs such as Nano introduce critical required servers, so 'fake tech' :-)

**Comment by Tim-W** [comment]
Integrated bitcoinJ library:
* In-app wallet screen to view your status, public key, funds
* Synchronizing the chain
* **From now on**, whenever a person submits a new album/EP/single, his/her wallet public key is added to the trustchain block, so that other users can send money. Therefore only **newly made releases after this 'protocol upgrade'** will have a tip/donate button
* Sending $1 tips to the artist directly from a playlist
* Direct conversion of USD/BTC as a helper for the user (obviously using an external API so this might be removed for "fully distributed" cleanness)

No real device-to-device bitcoin transactions tested yet, as BTC transaction fees are high, up for discussion tomorrow :P
<img src="https://user-images.githubusercontent.com/4957928/91071322-52354000-e638-11ea-8a8d-c21432c15658.gif" width="250"/>

**Comment by synctext** [comment]
Comments on thesis draft:
- "How can we design and implement a music streaming service that distributes the power from one authority
to its users?", congrats! that is a good research focus.

**Comment by Tim-W** [comment]
**Progress toward Remote Keyword Search**
I implemented a simple, fully distributed algorithm which uses neighbour nodes to find content. Basically, it works as follows:
*Scenario: a user wants to find content with keyword "Eminem"*
1. User checks its own local database (downloaded trustchain blocks) to filter out content by this keyword
2. If there are less than X results, ask neighbours (max N) to obtain it, by sending a "KeywordSearchMessage": 
```
query: "Eminem"
ttl: 3
origin: <my public key>
```
3. When a peer receives a KeywordSearchMessage: check local database for results.
If there are results: send results as trustchain blocks directly to original peer and terminate.
4. If ttl > 0:
Reduce ttl by 1, and forward message to own neighbours (max N). Otherwise terminate.

This is a simple (but effective?) algorithm for doing remote keyword search. I think it is a somewhat similar algorithm as used in [remote_query_community.py](https://github.com/Tribler/tribler/blob/devel/src/tribler-core/tribler_core/modules/metadata_store/community/remote_query_community.py), but simplified (no SQL). I implemented the code in Kotlin here: 
https://github.com/Tim-W/trustchain-superapp/commit/18ea0d1ae6833d00ffe23fa091d94a8165dd41b6 

It requires testing & device-to-device experimentation. Is this a step in the right direction for search?

**Comment by synctext** [comment]
Good stuff! Please simplify and always ask 20 peers around you.

**Comment by synctext** [comment]
ttl: 3
Actually, there is no need for Gnutella explosion of query traffic. Keep a local database and caching.

**Comment by Tim-W** [comment]
Searching.gif
![ezgif-6-bda4ac3b42e8](https://user-images.githubusercontent.com/4957928/92386838-d72e5800-f114-11ea-8515-0b85f92833f8.gif)
Donation.gif (please note that the donation is not yet a confirmed transaction after sending. On my remote server I have a simple miner script that mines 50 blocks every 10 seconds, to test transactions on my own testnet.)
![ezgif-6-0f2ca91611b4](https://user-images.githubusercontent.com/4957928/92386840-d8f81b80-f114-11ea-9e1e-0c6e92a2b867.gif)

**Comment by synctext** [comment]
Great progress!
- Field trail realistic
  - goal: show irrefutable results of you unique autonomous system
  - Experiment: We show that money flows without **any** middleman
  - A total of X Tweakers users found free money and used our Tip-Jar feature
- Question of user-generated content
  - [ ] offer free money to users in forked Bitcoin network
  - [X] support keyword search
  - [ ] support popularity community
  - [ ] support channels

**Comment by Tim-W** [comment]
Currently I am mostly working on implementation.tex (see recent updates to https://github.com/Tim-W/msc-thesis, report: https://github.com/Tim-W/msc-thesis/blob/master/report.pdf)

**Comment by synctext** [comment]
Reading latest [version (direct raw thesis .PDF)](https://github.com/Tim-W/msc-thesis/raw/master/report.pdf)

Discussed "DAO", what is proposed in this thesis on the Organisation side?
Brainstorm: all users contribute income for artists and get to vote, artists create content and get to vote also. Each Bitcoin either donated by a fan or received by an artist obtains proportional voting rights inside the DAO (or capped for abuse/central control purposes; voting threshold).

**Comment by Tim-W** [comment]
[report.pdf](https://github.com/Tim-W/msc-thesis/raw/master/report.pdf)
*Changes to the report*
- Rename related work->state of the art
- Add Tribler, TrustChain-Superapp to state of the art
- Add design choices (short explanation) to design for the use of TrustChain, TrustChain-Superapp, Release blocks model, Bitcoin.
- Make Implementation more readable by starting with showing the interface
- Add a shortlist of features to implementation
- Table of libraries usage of implementation
- Simplified many parts of implementation and make it more high-level
- Add QA section in implementation

*Changes to the code*
- Implemented more unit tests
- Upgraded dependencies

**Comment by synctext** [comment]
Reviewed current thesis draft, making solid progress. All content is there, just present it at a higher abstraction level:
- Content discovery, versus "artist identification"
- Tamper-proof storage, versus "trustchain interaction"
- Zero-Infrastructure communication, versus "phone-to-phone serverless communication"
- Section 4.1 Zero-Server architecture, versus "4.1 Phone-to-phone connectivity"
- Section 4.2 End-to-end ecosystem; from artists to fan ears
- Section 4.3 Grounding of an alternative industry; not perfect but future proof and permissionless infrastructure which can be improved upon by anyone

**Comment by Tim-W** [comment]
This week, mainly focused on making visualizations/models of current vs. desired situation in distributed vs. centralized music streaming services. Also added pseudocode for the remote keyword search function.  See problem description & design. I tried to make the design chapter a bit more abstract/high-level (mainly the first sentences and titles of sections).

Now that I filled the report with more models, I need to introduce more explanations in text, that is the next focus.

See [report.pdf](https://github.com/Tim-W/msc-thesis/blob/master/report.pdf)

3 Short questions:
* @synctext I use "We" everywhere, but did not add you yet as an author/contributor. How should I mark your name in the thesis?
* Is pseudocode for algorithms more logical in implementation? In the pseudocde I left implementation details so that's why I left it in design for now.
* Same Q. as above, for data models (see figs. 4.4 and 4.3: Release blocks and KeywordSearchMessage)

**Comment by synctext** [comment]
- There is a standard TUDelft template for thesis work, thesis committee members are listed there.
- pseudocode looks look there
- datamodels; Chapter implementation has too much details. If the reader wants to know: read the code.
- Could be more scientific. Thesis still can use more abstraction. (now more blogger terms in Figure 2.1) These figures could be more detailed. More abstract and business level: show a money table or 2%-ish [revenue cuts](https://www.orpheusaudioacademy.com/spotify-pay/) that go to the creator.

related work. Obviously this is false, as your MusicDAO work [pre-dates LikeCoin](https://docs.ipfs.io/concepts/case-study-likecoin/#the-story): "As the world's first data registry specifically recording the internet's digital creative content, it stores ISCN metadata for every item published or shared within Liker Land or associated with the LikeCoin button.",

**Comment by synctext** [comment]
More flavor for the "problem Description" at a high abstraction: business-level:
```
Wixen raised the case in December 2017, claiming that the music-tech giant was hosting tens of thousands of songs
managed by Wixen on its streaming service without holding the proper copyright to them. It also said Spotify offered
“outrageous annual salaries to its executives” while skimping on artist payments.
```
[Spotify Settles Its $1.6 Billion Publishing Lawsuit](https://www.rollingstone.com/pro/news/spotify-settles-its-1-6-billion-publishing-lawsuit-771557/)
```
The court papers accuse the streaming service of lacking the infrastructure needed to make sure songs are
licensed and musicians are paid.
```
[Spotify sued over 'billions of Eminem streams'](https://www.bbc.com/news/technology-49436077) by BBC
Bad IT in music biz, by Daan: https://www.copyrightdelta.com/media/CopyrightDelta-Presentation-CordaCon2020.mp4

**Comment by Tim-W** [comment]
Thank you! will look into it. There are a lot of different problems in the music industry right now, so I could go on forever in the problem description section.. So I will need to filter out until we have the most important!

https://github.com/Tim-W/msc-thesis/blob/master/report.pdf
Work from last days:

-  Added intro
-  Added to problem description: revenue cut table, subsection 2.4.2 
-  Rewritten most of state of the art and extended it to 4 pages
-  Design: Tried to reduce complexities. Added 'End-to-end music delivery model' (with figure 4.2). Added % to figures 4.5/4.6.
-  Implementation: Removed implementation/engineering details. Tried to make everything more readable/abstract (Identity/authenticity, priority handling). More explanations for the interfaces.

**Comment by synctext** [comment]
Lastest .PDF download link: https://github.com/Tim-W/msc-thesis/raw/master/report.pdf

Fresh idea that I also had: Big Tech platforms (Youtube, Netflix Apply music, Spotify) are already the first wave of the robot economy [ref](https://www.opendemocracy.net/en/can-europe-make-it/robot-economy-full-automation-work-future/). Feel free to include this within your thesis storyline.

**Comment by synctext** [comment]
## The first robot economy
Apologies for giving conflicting advice yet again. [April 6](https://github.com/Tribler/trustchain-superapp/issues/45#issuecomment-609681925) comment about the "first robot economy" also has great potential as superior scientific storyline. Because you did not implement any AI, I incorrectly assumed it wasn't a valid storyline.
![image](https://user-images.githubusercontent.com/325224/97419787-c6c37e00-190a-11eb-917f-ac2097d9fd71.png)

Please consider adding more of AI storyline in for the final thesis writing sprint. No need to have any AI within the implementation: you worked on the cardinal first step towards an AI DAO; deployed in the wild with zero-server architecture for a zero-cost music industry. Compared to music labels and platform overhead (e.g. 90% of cost) you can sustain 800% to 1000% increase in artists living of their work.

**Brainstorm, Chapter 3: Problem Description (leave the chapter as is, insert this up front and see how it goes)**
The central problem addressed within this thesis is exploring new architectures for the first "AI DAO" applied to a specific industry. A DAO is an autonomous organization with a decentralized governance, budgeting system and without traditional employees or managers. An "AI DAO" is the envisioned idea of a leaderless organisation with a decision making process controlled by AI. Our goal is to create the first AI DAO framework and deploy this in the wild to further evolve. We will focus on a viable framework itself within this thesis and leave the increase in intelligence and learning to future work.

For many decades AI scientists have failed to materialise in the wild the more basic concept of an "intelligent Agent". This is an "Autonomous entity that acts, directing its activity towards achieving goals, upon an environment using observation through sensors and consequent actuators. Intelligent agents may also learn or use knowledge to achieve their goals."

Much is still unknown within the young AI DAO field. We do not yet understand on how to achieve them or give them intelligence. Especially governance issues are still insurmountable, who owns them and who ultimately controls them? What even makes something autonomous in the first place?

This thesis acts as a first concrete step to increasing our understanding of AI DAO. By trying to engineer an AI DAO we will irrefutably identify the gap between theoretical speculations and running code. This empirical approach leads us to the selection problem of which business to automate with our AI DAO framework.

We now present an analysis of the music industry because we believe for three reasons this is the most fertile starting point for our AI DAO framework called "TimDAO". First, the music industry has an overhead percentage of 90%. Only 10% of the revenue stream goes to the producers of the core product: music artists. Second, this industry is inherently digital. Digital music recording and digital music playback ensure that the entire industry, _in theory_, could be replaced by TimDAO. Third, due to [anti-competitive practices such as price fixing ](https://scholar.google.com/scholar?q=price+fixing+music+industry) this industry has lost the moral high ground. Fourth, the music industry value chain is mature and numerous components have already been digitised using open source libraries. We can re-use many building block to implement the TimDAO framework and realistically release it in the wild.
 {_expand the above into 2 pages with existing [fantasy blog posts](https://towardsdatascience.com/why-building-an-ai-decentralized-autonomous-organization-ai-dao-85d018700e1a) and thoughtful ideas on [AI DAO](https://blog.ethereum.org/2014/05/06/daos-dacs-das-and-more-an-incomplete-terminology-guide/) as inspiration_}
![image](https://user-images.githubusercontent.com/325224/97425168-0477d500-1912-11eb-9783-409539d24746.png)

**Comment by Tim-W** [comment]
Thanks for the good resources.

`*Brainstorm, Chapter 3: Problem Description (leave the chapter as is, insert this up front and see how it goes)*`
`Keep Chapter 3 simple: economic literature on centralisation of power; Section 3.2 Alternative music industry initiatives (organisational!); Subsection "3.3 Decentralised technology alternatives" (Good Kind of Tech) with IPFS, DAO, Superapp matters.`

Currently, Chapter 3 has title "State of the Art", but with the above comments it seems that it should include a Problem Description as well? `economic literature on centralisation of power` seems more logical to be in Problem Description (Chapter 2) to me. Then with Chapter 3 listing all current decentralized organizational structures (non-tech), then DAO applications/protocols outside of music industry (tech) and then DAO applications/protocols inside the music industry (tech). Does that make more sense?

I will try to implement this AI DAO storyline, but I am not so sure if it is realistic to incorporate a new central element in my thesis, with deadlines growing closer.

**Comment by Tim-W** [comment]
@synctext Could you check my email? Although I'm unsure if 18th of December is still viable as thesis date as my progress is not going that quickly as I wanted to

**Comment by synctext** [comment]
Latest [.PDF download of thesis](https://github.com/Tim-W/msc-thesis/raw/master/report.pdf), remarks:
- Putting AI DAO as central storyline is going well
- Thesis is maturing and progressing nicely
- 4.2.1. Rename 'Autonomy and resilience' was "Censorship-resiliency"
- 5.1. Zero-server infrastructure (moved forward)
- 5.1. Zero-server music industry (rewording of app features)

**Comment by Tim-W** [comment]
1. Created bitcoin faucet: every user that obtains a fresh wallet key, sends this key to my faucet server, and receives 10 fake dollars.
2. Created an IPv8 message type for sending swarm health messages. Basically implemented most stuff of popularity_community.py (from Tribler) in Kotlin.
3. Devices now keep track of swarm health of 5 popular torrents and 5 random torrents internally.
4. Devices once every X seconds send a swarm health message to a random peer (this needs testing and some work to finish up the end-to-end functionalities)
4. Album overview is sorted based on swarm health.
5. Fix some bugs: Music keeps playing while browsing, syncing wallet blockchains, downloading/buffering progress bars

Will post some screenshots/GIFs when I have the popularitycommunity end-to-end working. APK on Nov 26

**Comment by synctext** [comment]
[MPs to investigate whether artists are paid fairly for streaming music](https://www.bbc.com/news/entertainment-arts-54551342)

**Comment by ichorid** [comment]
Tried to install the abovementioned `app-debub.zip` on my phone, the phone says "syntactic error while opening the packet". What am I doing wrong?

p.s. Feel free to delete this message if this is off-topic.

**Comment by Tim-W** [comment]
@ichorid Please try using this artifact https://github.com/Tribler/trustchain-superapp/suites/1636262534/artifacts/30239452 let me know if there are issues. Unpack the zip on your phone -> open the APK.

**Comment by Tim-W** [comment]
Added cover art & new layout (nothing fancy, if the album is pre-downloaded it fetches the cover art from the phone cache, no fetching from other devices directly).
<details>
<summary>Cover art & new layout</summary>
<img src="https://user-images.githubusercontent.com/4957928/101764253-23a47d80-3ae0-11eb-9ff5-31d2b5056664.png" width=200/>
</details>
I'm currently in the progress of obtaining 100+ creative common albums and making torrents and wallets for the demo. Then they can be put on seed box.

**Comment by synctext** [comment]
[.APK downloading link fails](https://github.com/Tribler/trustchain-superapp/suites/1636262534/artifacts/30239452) to work on Android
![Screenshot_20201214_155250_com fastaccess github](https://user-images.githubusercontent.com/325224/102096107-a13ff480-3e24-11eb-8478-6ebd51e0b1a9.jpg)

**Comment by Tim-W** [comment]
https://github.com/Tribler/trustchain-superapp/suites/1665275830/artifacts/30996859

**Comment by Tim-W** [comment]
[app-14dec.zip](https://github.com/Tribler/trustchain-superapp/files/5689656/app-14dec.zip)

**Comment by devos50** [comment]
It seems that Spotify is supporting direct artist donations without taking a cut:

<img width="414" alt="Schermafbeelding 2020-12-21 om 09 54 04" src="https://user-images.githubusercontent.com/1707075/102758233-b9f36180-4372-11eb-9a01-bcbbaff5829b.png">

**Comment by synctext** [comment]
Play store version works! Donated 1BTC to a good free album.

**Comment by synctext** [comment]
btw direct donation with Spotify still have PayPal as an intermediary taking a cut.

**Comment by Tim-W** [comment]
**MusicDAO release 0.2 "Christmas release"**
Google Play store: https://play.google.com/store/apps/details?id=nl.tudelft.trustchain&gl=NL
**The app is ready for its release experiment.** I will send a message over Slack to other 'colleagues' to download, install and run the MusicDAO.

I currently have 2 "seedbox" nodes running to get the music data started. There are a little over 100 albums published on TrustChain with torrents, with 3.9GB of music. I have 20 virtual IPv8 nodes running which are publishing this data, spread over 2 servers (10 nodes each).
I also renamed "Ipv8 apps" to "TrustChain Superapp" for more consistent naming. 

Latest new features and changes https://github.com/Tribler/trustchain-superapp/pull/50 :
* Various stability improvements and fixes
* Drag to refresh overview
* Streaming implementation improvements
* Fixed inconsistent naming of track files

<details>
<summary>GIF: Browsing and streaming music</summary>
<img src="https://user-images.githubusercontent.com/4957928/103173843-e5d48280-485d-11eb-907a-e20c9b25fbba.gif" width=250/>
</details>
<details>
<summary>GIF: Seinding money to artists</summary>
<img src="https://user-images.githubusercontent.com/4957928/103173848-e836dc80-485d-11eb-9863-660601a5ddf7.gif" width=250/>
</details>

Main challenges currently:
* Downloading track lists (metadata) is usually very quick, although downloading the MP3s can take much longer. The bittorrent peers are connecting well but throughput is often 0.0. Need to investigate more what's going on here.
* IPv8 nodes eating up memory of the servers
* Possible bugs

**Comment by synctext** [comment]
- Status of master thesis
  - Now bug fixing, 100% of the code is complete
  - Thesis writing, 75% done.
- DHT announce from 1 IPv4 within IPv8 for 100 swarms
  - probably detected as a DDoS attack by the other participants in DHT
  - Our Ipv4 is banned then
  - results in extremely slow content discovery speed
  - Performance analysis: without DDoS fix, with central tracker, and with DDoS fix by using multiple IPv4 addresses
- Quality assurance content
  - Figure 5.9: Firebase Crashlytics crash reporter; this figure shows the crashes over time
  - Document the bugfixes

**Comment by Tim-W** [comment]
Finally, after getting some rid of some pesky bugs in the seedboxes, I'm creating the first graphs measuring performance. Lot of work in evaluation/measurement still to do
<img src="https://user-images.githubusercontent.com/4957928/105528235-040a9280-5ce5-11eb-866a-457c6c5d2422.png" width=400/>
Also I am now able to 'control' a network of 6 real world Android devices, for doing experiments. (and for making demo videos for the thesis defence?)
<img src="https://user-images.githubusercontent.com/4957928/105530064-957b0400-5ce7-11eb-96f0-051e9c4ff33b.jpg" width=400/>

**Comment by synctext** [comment]
That graph is good thesis stuff!

**Comment by Tim-W** [comment]
Started with measurements/evaluation: now I have 4 graphs and a tiny bit of text: 
[evaluation-v1.pdf](https://github.com/Tribler/trustchain-superapp/files/5870322/evaluation-v1.pdf)

Still to do: - graph/list of all transactions over time (non related to faucet transactions) and some nice visualization showing which artists received money, and when this was transacted. 
- many other graphs/evaluation material

Last 2 weeks:
* Fixed various bugs, new releases on google play store
* Finally fixed seedboxes well, downloading an album should be 10-20 seconds now
* Ran experiments with ~5 android devices locally (see pdf)
* Gave the MusicDAO to 10 people so far, to let them play around with it and send money to artists. I can inspect this money flow by inspecting the blockchain

**Comment by synctext** [comment]
- Good chapter 7 matter.
- minor quality of life improvement: click triggers download of swarm, also auto-play please when 50% downloaded or 100% downloaded.
- Goal of this first experiment is to determine the responsiveness of our infrastructure in ideal circumstances. We scripted our downloading and streaming activities to determine the lower bound for the end-to-end latency. This experiment uses hardware, thus Android will be less. etc. 
- "we inspected the blocks in our Bitcoin regtest environment in order to show all transactions", re-formulate into more scientific focus. like: We obtained a dataset for our experiments by combining information from various sources .. combining the Google Play transactions with Bitcoin monitoring, etc.

**Comment by Tim-W** [comment]
New graphs, that show the transaction success of our bitcoin blockchain environment.
<details>
<summary>(Cumulative) artist income vs. bitcoin block count</summary>
<img src="https://user-images.githubusercontent.com/4957928/106151634-d7d79180-617c-11eb-931c-f86cf9edd38e.png"/>
</details>
<details>
<summary>(Cumulative) transactions vs. bitcoin block count</summary>
<img src="https://user-images.githubusercontent.com/4957928/106151640-d9a15500-617c-11eb-8533-c9d022d12dd2.png"/>
</details>

**Comment by Tim-W** [comment]
**Update 10/02/2021 created MusicDAO demo video**
Currently 1:27 minutes, could remove some stuff. It is not very professional, but it does show all its functionalities in quite a short time.


https://user-images.githubusercontent.com/4957928/107580497-edc26900-6bf6-11eb-8c82-2f83d6cdf704.mp4

**Comment by synctext** [comment]
- Learnings: running 6 Android emulators on local laptop as ideal development environment. (plus 3 real phones and unit tests
- 9 crashes reported inside Kotlin-IPv8 messaging since 5Feb; 7 other crashes
  - create issue(s) with details please (fixing is something else, essential to know pending quality issues)
  - 2 inside Trustchain payload generator
- Video is great intro to your thesis work and publicity for user-trail + measurements of live network.
  - users may be confused where music comes from and what the copyright status is.
  - suggested text to add "Streaming any Bittorrent audio swarm", "Grow the Creative Commons"
- Thesis review, 48 page version.
- Table 5.2: Code coverage overviewTable 5.2: Code coverage overview: please add "Total Lines of Code" for all your work.
- Figure 6.4: Average time spent per download stage; cleaner: make it a graph going to 120 seconds.
- 5.11. Quality assurance: "In order to preserve quality of code we make use of continuous integration", claim credit for all the nice work you did. like 'Quality assurance is particular important aspect of this thesis work. With high quality code we facilitate potential re-usage of our ideas and implementation within industries outside the music business. We used industry best-practices such as continous integration'... 'Finally we used 20 unique Android smartphones for hands-on bug mitigation, User Experience uniformity, and performance analysis. Table X lists the various... During the implementation phase we ensured that MusicDAO operates smoothly on all these different Android versions, screens, and processors.'

**Comment by Tim-W** [comment]
<details>
<summary>I got a second bootstrap server succesfully running thanks to Sandip.</summary>
<img src="https://user-images.githubusercontent.com/4957928/107977720-dc44dc80-6fbb-11eb-8b56-4920926d3a00.png" width=250/>
</details>
Reduced the music gossip time for a few devices; 5 Androids are now sending music blocks every 0.5 seconds (on day&night).

An actual end-to-end video of peer-to-peer libtorrent and bitcoin payment! The phone on the left is the only uploader and published the "John Doe" album. The tablet consumes it and sends bitcoins to the phone. See here: https://drive.google.com/file/d/1XKU91P0lShjQHtztdCMpFC6U4sAy-cfx/view?usp=sharing Will record a better version of this for the final defense presentation.

**Comment by synctext** [comment]
Nice!, please push speed (initial music discovery) and robustness (bootstraping) improvements  to Android store.
today if you can.

**Comment by Tim-W** [comment]
* Started on a more coherent Introduction and main storyline (theory around robot economy still needs some work): 

Problem of centralization/platformization on the Internet > What is a robot economy (theory) -> What is a robot economy in software? -> What does it achieve (key characteristics) -> Problem of centralization in music industry -> Our first steps towards robot economy: MusicDAO

* Added conclusion + generalization of our framework
* Started on Future work (will be very long, so should this be a separate chapter?)
* Reduced focus on censorship-resilience
* Evaluation: P2P bootstrap problem section
* Evaluation: Battery usage section
* Evaluation: Scalability
* Evaluation: improved graph Content discovery and fixed maths
* Implementation: improved quality assurance section
* Design: add content popularity gossiping algorithm
* Design: add centralized vs. phone-to-phone netwokr diagrams
* Overall: Improve layout of images
* Overall: Improve layout of sections, reduce fragmentation of sub(sub)sections
* Problem description: added financial transparency section
* Problem description: added figure of music revenue over time in different sectors
[report-20-02-2021.pdf](https://github.com/Tribler/trustchain-superapp/files/6014902/report-20-02-2021.pdf)

**Comment by synctext** [comment]
quick remarks, that robot economy architecture picture is great! Please make this the centerpiece of your solution and final thesis presentation. _This is the future of the firm!_ Spend a few hours to make it stand-out if possible:
- remove the distributed data store
- "Collective Money" as separate box :ballot_box: 
- give it an aesthetically pleasing star-shape with DAO at center with 4 connecting items (users,creators,robo infra,money)
- polished icons for everything, like open source: <img src="https://user-images.githubusercontent.com/325224/109127083-4abd3380-774e-11eb-9872-a9929e40f4d2.png" width=50>
- Icon for AI, single icon for user, icon for creative type/artist :art: , green icon for money :dollar: (or use the €10 banknote which is red :euro:)

**Comment by Tim-W** [comment]
<details>
<summary>More updates week 2 of writing</summary>
* Overall: made all sections roughly the same length (if they were too long, added subsections)
* Improved robot economy image, improved its storyline in connection to 3.2 "AI+DAO" and Design intro
* Implementation: add Content Popularity gossip algorithm
* Implementation: improve keyword search algorithm sec + add screenshots
* Implementation: make introduction more clear
</summary>
[MSc_Thesis_25-02-2021.pdf](https://github.com/Tribler/trustchain-superapp/files/6044714/MSc_Thesis_25-02-2021.pdf)

**Comment by synctext** [comment]
- "Democratic user engagement" suggest not negative cross, but 'out of scope'.
- citation style, (Rayna & Striukova, 2009), switch to "[9]" if not too much work.

**Comment by Tim-W** [comment]
Re-done n=10 content discovery experiment as the previous run suffered from network congestion
![expected-vs-simulated-releases-2](https://user-images.githubusercontent.com/4957928/110515548-cb920d00-8108-11eb-9abe-d186c9b14773.png)
Re-done search latency experiment, now timed without UI rendering time
![search-latency-2](https://user-images.githubusercontent.com/4957928/110515553-cd5bd080-8108-11eb-911c-96355a022133.png)
